console-subscriber/src/record.rs

```diff
@@ -0,0 +1,224 @@
+
+impl Recorder {
+    pub(crate) fn new(path: &Path) -> io::Result<Self> {
+        let buf = Arc::new(Mutex::new(RecordBuf::new()));
+        let buf2 = buf.clone();
+        let file = std::fs::File::create(path)?;
+
+        let worker = std::thread::Builder::new()
+            .name("console/subscriber/recorder/io".into())
+            .spawn(move || {
+                record_io(Io { buf: buf2, file });
+            })?;
+
+        let recorder = Recorder { buf, worker };
+
+        recorder.write(&Header {
+            v: DATA_FORMAT_VERSION,
+        });
+
+        Ok(recorder)
+    }
+
+    pub(crate) fn record(&self, event: &crate::Event) {
+        let event = match event {
+            crate::Event::Spawn { id, at, fields, .. } => Event::Spawn {
+                id: id.into_u64(),
+                at: *at,
+                fields: SerializeFields(fields),
+            },
+            crate::Event::Enter { id, at } => Event::Enter {
+                id: id.into_u64(),
+                at: *at,
+            },
+            crate::Event::Exit { id, at } => Event::Exit {
+                id: id.into_u64(),
+                at: *at,
+            },
+            crate::Event::Close { id, at } => Event::Close {
+                id: id.into_u64(),
+                at: *at,
+            },
+            crate::Event::Waker { id, op, at } => Event::Waker {
+                id: id.into_u64(),
+                at: *at,
+                op: *op,
+            },
+            _ => return,
+        };
+
+        self.write(&event);
+    }
+
+    fn write<T: Serialize>(&self, val: &T) {
+        let mut buf = self.buf.lock().unwrap();
+        serde_json::to_writer(&mut buf.bytes, val).expect("json");
+        buf.bytes.push(b'\n');
+        drop(buf);
+        self.worker.thread().unpark();
+    }
+}
+
+impl RecordBuf {
+    fn new() -> Self {
+        Self {
+            bytes: Vec::new(),
+            next: Vec::new(),
+        }
+    }
+
+    /// Takes the existing bytes to be written, and resets self so that
+    /// it may continue to buffer events.
+    fn take(&mut self) -> Vec<u8> {
+        let next = std::mem::take(&mut self.next);
+        std::mem::replace(&mut self.bytes, next)
+    }
+
+    fn put(&mut self, mut next: Vec<u8>) {
+        debug_assert_eq!(self.next.capacity(), 0);
+        next.clear();
+        self.next = next;
+    }
+}
+
+fn record_io(mut dst: Io) {
+    use std::io::Write;
+
+    loop {
+        std::thread::park();
+
+        // Only lock the mutex to take the bytes out. The file write could
+        // take a relatively long time, and we don't want to be blocking
+        // the serialization end holding this lock.
+        let bytes = dst.buf.lock().unwrap().take();
+        match dst.file.write_all(&bytes) {
+            Ok(()) => {
+                dst.buf.lock().unwrap().put(bytes);
+            }
+            Err(_e) => {
+                // TODO: what to do if file error?
+            }
+        }
+    }
```

Comment:
we probably want some way to ensure that the file is flushed when the program is shut down (e.g. if it panics, we don't want to miss the last event).

but, we can probably do that in a follow-up...